pyanaconda/subscription.py

```diff
@@ -0,0 +1,178 @@
+# - from UI based on user interaction
+#
+# Also in some cases, multiple individual DBus tasks will need to be run.
+# Anaconda modularity on RHEL8 is not advanced enough to handle this by itself,
+# so we need s imple sheduler living in the context of the main Anaconda
+# thread, that hosts the code that starts the respective subscription thread,
+# that makes sure appropriate tasks are run.
+#
+# This code than can be run either during early startup or in reaction to user
+# interaction in the UI, avoiding code duplication.
+
+from enum import Enum
+
+from pyanaconda.threading import threadMgr
+
+from pyanaconda.core.i18n import _
+
+from pyanaconda.core.constants import THREAD_WAIT_FOR_CONNECTING_NM
+from pyanaconda.core.constants import RHSM_AUTH_USERNAME_PASSWORD, RHSM_AUTH_ORG_KEY
+
+from pyanaconda.modules.common.constants.services import SUBSCRIPTION
+from pyanaconda.modules.common.task import sync_run_task
+
+
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+
+class SubscriptionPhase(Enum):
+    UNREGISTER = 1
+    REGISTER = 2
+    ATTACH = 3
+    DONE = 4
+
+def dummy_progress_callback(phase):
+    """Dummy progress reporting function used if no custom function is set."""
+    pass
+
+def dummy_error_callback(error_message):
+    """Dummy error reporting function used if no custom function is set."""
+    pass
+
+def subscribe(progress_callback=None, error_callback=None):
+    """Try to register a system."""
+
+    # assign dummy callback functions if none were provided
+    if progress_callback is None:
+        progress_callback = dummy_progress_callback
+    if error_callback is None:
+        error_callback = dummy_error_callback
+
+    error_message = ""
+    # connect to the Subscription DBus module
+    subscription_proxy = SUBSCRIPTION.get_proxy()
+
+    # first sure network connectivity is available
+    threadMgr.wait(THREAD_WAIT_FOR_CONNECTING_NM)
+
+    # Then check if we are not already registered.
+    #
+    # In some fairly bizzare cases it is apparently
+    # possible that registration & attach will succeed,
+    # but the attached subcription will be uncomplete
+    # and/or invalid. These cases will be caught by
+    # the subscription token check and marked
+    # as failed by Anaconda.
+    #
+    # It is also possible that reigstration succeeds,
+    # but attach fails.
+    #
+    # To make recovery and another regsitration attempt
+    # possible, we need to first unregister the already
+    # registered system, as a registration attempt on
+    # an already registered system would fail.
+    if subscription_proxy.IsRegistered:
+        log.debug("RHSM: subscription thread: system already registered, unregistering")
+        progress_callback(SubscriptionPhase.UNREGISTER)
+        task_path = subscription_proxy.UnregisterWithTask()
+        task_proxy = SUBSCRIPTION.get_proxy(task_path)
+        sync_run_task(task_proxy)
+        error = task_proxy.Error
```

Comment:
This is just wrong. You should handle an exception.

Reply:
OK, lets do that in the context of the error handling improvements PR.